program/rust/src/tests/test_permission_migration.rs

```diff
@@ -0,0 +1,268 @@
+use crate::c_oracle_header::{
+    MappingAccount,
+    PermissionAccount,
+    PriceAccount,
+    ProductAccount,
+    PC_VERSION,
+};
+use crate::deserialize::initialize_pyth_account_checked;
+use crate::error::OracleError;
+use crate::instruction::OracleCommand::{
+    AddMapping,
+    AddPrice,
+    AddProduct,
+    AddPublisher,
+    DelPrice,
+    DelProduct,
+    DelPublisher,
+    InitMapping,
+    InitPrice,
+    ResizePriceAccount,
+    SetMinPub,
+    UpdProduct,
+};
+use crate::instruction::{
+    AddPriceArgs,
+    AddPublisherArgs,
+    CommandHeader,
+    DelPublisherArgs,
+    InitPriceArgs,
+    SetMinPubArgs,
+};
+use crate::processor::process_instruction;
+use crate::tests::test_utils::AccountSetup;
+use bytemuck::bytes_of;
+use solana_program::pubkey::Pubkey;
+
+#[test]
+fn test_permission_migration() {
+    let program_id = Pubkey::new_unique();
+
+    let mut permissions_setup = AccountSetup::new_permission(&program_id);
+    let permissions_account = permissions_setup.to_account_info();
+
+    let mut funding_setup = AccountSetup::new_funding();
+    let funding_account = funding_setup.to_account_info();
+
+    let mut attacker_setup = AccountSetup::new_funding();
+    let attacker_account = attacker_setup.to_account_info();
+
+    let mut mapping_setup = AccountSetup::new::<MappingAccount>(&program_id);
+    let mut mapping_account = mapping_setup.to_account_info();
+
+    let mut next_mapping_setup = AccountSetup::new::<MappingAccount>(&program_id);
+    let mut next_mapping_account = next_mapping_setup.to_account_info();
+
+    let mut product_setup = AccountSetup::new::<ProductAccount>(&program_id);
+    let mut product_account = product_setup.to_account_info();
+
+    let mut price_setup = AccountSetup::new::<PriceAccount>(&program_id);
+    let mut price_account = price_setup.to_account_info();
+
+
+    product_account.is_signer = false;
+    mapping_account.is_signer = false;
+    price_account.is_signer = false;
+    next_mapping_account.is_signer = false;
+
+
+    {
+        let mut permissions_account_data =
+            initialize_pyth_account_checked::<PermissionAccount>(&permissions_account, PC_VERSION)
+                .unwrap();
+        permissions_account_data.master_authority = *funding_account.key;
+    }
+
+    assert_eq!(
+        process_instruction(
+            &program_id,
+            &[
+                attacker_account.clone(),
+                mapping_account.clone(),
+                next_mapping_account.clone(),
+                permissions_account.clone()
+            ],
+            bytes_of::<CommandHeader>(&AddMapping.into())
+        ),
+        Err(OracleError::PermissionViolation.into())
+    );
+
+
+    process_instruction(
+        &program_id,
+        &[
+            funding_account.clone(),
+            mapping_account.clone(),
+            permissions_account.clone(),
+        ],
+        bytes_of::<CommandHeader>(&InitMapping.into()),
+    )
+    .unwrap();
+
+    assert_eq!(
```

Comment:
I'm confused about the tests below. I thought that `funding_account` is permissioned to do these actions, and the `attacker_account` isn't. It looks like every test below is a permission violation

Maybe leave some code comments explaining what the tests are supposed to be testing.